### PR TITLE
Add basic message formatting + other stuff

### PR DESCRIPTION
**Pull request by heyitsmeuralex**
* client actually looks nice now
* usernames must satisfy `[a-zA-Z0-9_-]+` regex
* `code` and \@mentions are now formatted
* \@mentions to everyone or the current user are highlighted
* because the client doesn't have a list of users it simply formats all valid user mentions regardless of whether the user actually exists or not -- fix this!

image

### CODE REVIEW

**Comment by heyitsmeuralex** [comment]
image

**Comment by heyitsmeuralex** [comment]
image

**Comment by towerofnix** [comment]
> $$latex$$

COOL AND GOOD

File: index.js [hunk 1]
```diff
@@ -185,6 +185,12 @@ async function main() {
     }
   })
 
+  app.post('/api/whoami', async (request, response) => {
+    const { username } = await getUserBySessionID(request.body.sessionID)
+
+    response.end(JSON.stringify({ username }))
+  })
+
```
Comment (towerofnix) [inline index.js]: Is this used in your PR?

It looks like you use it to get the name of the user from a session, which you use for highlighting mentions. You should use the new session API for this - or, better yet, just merge with my commits and use `(sessionObj && sessionObj.user.username === 'whatever')`.

File: index.js [hunk 2]
```diff
@@ -213,12 +219,20 @@ async function main() {
...
       return
     }
 
+    if (!reValidUsername.test(username)) {
+      response.end(JSON.stringify({
+        error: 'username invalid'
+      }))
+      return
+    }
+
```
Comment (towerofnix) [inline index.js]: Cool. I'll nitpick a couple things - this should have a status code (probably 400, Bad Request), and there should be a line break before `return`. (I know I'm being nitpicky, and that the changes I'm requesting are ones I made to my own code probably after you made this PR - sorry :) )

File: site/script.js [hunk 3]
```diff
@@ -12,6 +12,7 @@ const main = async function() {
   const socket = io()
 
   let sessionID
+  let user = null
```
Comment (towerofnix) [inline site/script.js]: Remove this in favor of the `sessionObj` object I added in recent commits.

File: site/script.js [hunk 4]
```diff
@@ -24,6 +25,10 @@ const main = async function() {
   if ('sessionID' in localStorage) {
     sessionID = localStorage.sessionID
     console.log('loaded session ID from local storage')
+
+    const { username } = await apiPost('/api/whoami', { sessionID })
+    user = { username }
+    console.log('username:', username)
```
Comment (towerofnix) [inline site/script.js]: Same deal here - use `(sessionData && sessionData.user.username === blah)`.

File: site/script.js [hunk 5]
```diff
@@ -169,6 +174,8 @@ const main = async function() {
     sessionID = result.sessionID
     localStorage.sessionID = sessionID
 
+    user = { username }
+
```
Comment (towerofnix) [inline site/script.js]: Also remove this.

File: site/script.js [hunk 6]
```diff
@@ -204,6 +211,70 @@ const main = async function() {
...
+        else if (char === '`' && currentToken === 'code') startToken('text')
+      }
+
+      buffer += char
+    }
+
+    startToken(null)
+    return el
+  }
+
```
Comment (towerofnix) [inline site/script.js]: Cool stuff! :tada:

File: site/styles.css [hunk 7]
```diff
@@ -37,6 +37,20 @@ body, html, #app {
...
+.message-mention-of-user {
+  color: white;
+  background-color: #cfa041;
+}
+
+.message-inline-code {
+  color: white;
+  background-color: #232323;
+}
+
```
Comment (towerofnix) [inline site/styles.css]: I think I prefer using the class `mention` and then using the CSS target `.message .mention`, but whatever - that's just me being pedantic about my own style.

**Comment by towerofnix** [comment]
Aaaaa I commented on an old commit. Show outdated please? Sorrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrry.

File: index.js [hunk 8]
```diff
@@ -290,6 +298,11 @@ async function main() {
       return
     }
 
+    // Don't give the following away, even to the user themselves.
+    delete user.passwordHash
+    delete user.salt
+    delete user._id
+
```
Comment (towerofnix) [inline index.js]: Why not?

File: index.js (hunk 2 above)
Comment (heyitsmeuralex) [reply index.js]: >the changes I'm requesting are ones I made to my own code probably after you made this PR - sorry

haha yes

**Comment by heyitsmeuralex** [comment]
Made changes as requested and merged with `master` over on this repo.
Also gg with making loads of comments on an old commit. I used your `sessionObj` thingy 😜

**Pull request merged**